Some changes in studies, redesign of options and arrows (issue 130 an…

olleeriksson opened the pull request:
Alright, here is most of the work from this weekend. I tried to split it up into multiple PR's but it turned out to be trickier than I had expected, so I am sorry about that. I will try to be more disciplined in any future PRs. This implements all the changes from both issue 130 and 132, so pretty much everything we have talked about.

To summarize:

- **Redesign of the options section.** The options are now below the chess board.
- **All arrows in a position are now shown, or none at all**, compared to before when only arrows for moves that hadn't been played more than two times were shown.
- **The option for showing arrows is now a multi-value option**, with two options that show arrows until all moves in the position have been played 2 times, and 5 times, as well as always showing all arrows, and never showing them.
- **A "Show hints for this move!" link has been added** that shows all arrows regardless of the currently selected arrow option. It has been given a slightly larger click area to be easy to hit quickly with the mouse or a finger on a mobile phone.
- **A "Reset line" option has been added**, which takes the user back to the starting position.
- **Renaming some options**, ie "Stockfish", "Favorite", "Show/hide arrows" "Slow/fast board reset", and "Key move" have been given slightly different names to be more clear.
- **Very minor issue with comments taking up an extra lines was fixed.** They are now trimmed of whitespaces and that extra empty is removed, gaining some space.

Note 1)  I just realized that there is a value limit of 5 in the tree. I hadn't seen that before. That kind of made my code for showing neglected moves somewhat moot. But I think it fulfills a purpose anyway even if it's just until move 5.

Note 2)  I used the mix gettext.extract --merge command to sync the pot/po files, but only checked in the study-files. The new version of this tool adds a bunch of elixir-autogen comments, and moves the existing comments one line, and updates the lines numbers obviously. So it looks complicated but viewed in a different diff viewer than GitHub perhaps there aren't really that many changes. But I think this was a safer way to do it than manually editing the files. And everything seems to work as far as I can tell.

Note 3)  I figured you are best suited to go into the german file and just search for msgstr "" and fill in those strings that don't have a translation. If you want to I could fill in the rest using Google Translate. You tell me.

Note 4)  There appears to be a bug when resetting progress where it doesn't always reset the values in the tree. I couldn't figure it out but might write an issue on it for later.

olleeriksson committed:
Some changes in studies, redesign of options and arrows (issue 130 and 132)

- Options below the chess board.
- All arrows in a position are now shown, or none at all.
- Arrow option is now a multi-value option.
- A "Show hints for this move!" option has been added.
- A "Reset line" option has been added.
- Renamed some options.
- Very minor issue with comments taking up space.

ArneVogel commented:
On note 3: please don't add automatic translations.
 
---

Some things I noticed:

When creating a study the initial position does not show the hint arrow. To reproduce just create a new study e.g. https://lichess.org/study/8q6X3jgB when it loads (or you refresh) it won't show an arrow from d2 to d4.

Toggling the setting changes their width. I would prefer them to stay the same width.
[before.webm](https://user-images.githubusercontent.com/26939450/203130406-ef6076c3-e98e-4722-a193-a952b7ce455f.webm)
[after.webm](https://user-images.githubusercontent.com/26939450/203130849-18ec7e7e-8c42-4869-9521-dd46ea133c02.webm)

e.g.
```
.study_option_item_below > a {
	display: block;
}
```
And each option having some specific `width: 23ch;` set.

olleeriksson committed:
Removed the auto-translated text strings.

olleeriksson committed:
Fixed some review comments

Removed autotranslated texts, and re-introduced the ground.redrawAll()
workaround to fix a problem with the arrows not showing up.

olleeriksson commented:
> On note 3: please don't add automatic translations.

I have removed them now!

> When creating a study the initial position does not show the hint arrow.

Hmm. Actually this turned out to be the same bug that we had the ground.redrawAll() workaround for before. The bug re-appeared when this line was removed so I re-added it. I spent a good few hours on it and still couldn't figure out why this is needed. It must be a bug in chessground. 

Anyway, I found a better place for this line though. Instead of putting it in give_hints() where it gets called every time you change the arrow settings or make a move, I put it in start_training() instead, where it has the same effect but doesn't bother the user anywhere near as much. Pretty happy with that solution.

> Toggling the setting changes their width. I would prefer them to stay the same width.

How about this. I set a min-width on all of them to the same value. That way if the screen is made smaller so that only 2 options can fit in one page it still looks ok. And if it happens that one translation happens to exceed the available space (difficult to know before every single translation has been done), then it just pushes the next option out a bit. There is no jumping back and forth now when toggling the options.

I could have made them slightly longer but I think having too much space between them looks weird too. And I don't think it fits 4 options per line.

Normal:
![image](https://user-images.githubusercontent.com/41130048/203167656-c0b4a2dd-ea32-4e98-82a4-535471e48aef.png)

Slightly narrower:
![image](https://user-images.githubusercontent.com/41130048/203167805-f362b819-3c02-4c4f-9f36-e5703d2f0d5a.png)

If one translation happens to be much longer:
![image](https://user-images.githubusercontent.com/41130048/203167883-16ab09e7-7ac2-4b8c-a064-a5c3b8b2a931.png)

olleeriksson commented:
By the way, a question not relating to this PR. What do you say about introducing an another option for "display opposite side's comments" which is on by default to cater for those users who have gotten used to it?

ArneVogel commented:
> I set a min-width on all of them to the same value.

Looks good

> What do you say about introducing an another option for "display opposite side's comments" which is on by default to cater for those users who have gotten used to it?

Now that there is more space for options I am not opposed to the idea.

ArneVogel commented:
Do you want to add the other ideas like storing options to localstorage to this pr or should I merge?

olleeriksson commented:
You can go ahead and merge!

I think I'm going to create separate PR on three things I've prepared.
- A link to Lichess analysis board.
- An option to enable/disable opposite side comments.
- Storing the options in local storage.

Would you prefer one or three PRs? One is probably a little easier, but I've prepared for making three this time if you prefer.

Maybe you want to add some icon for analysis board link btw? We could also change the icon for the Play against Stockfish if you want to while we're at it. I've circled some that might be handy to have below. The lower right is the Lichess symbol for a Study, which I thought could come in handy:

![image](https://user-images.githubusercontent.com/41130048/203408443-2696f7a4-367d-4bf0-ba2f-5a1bac760bee.png)

Also, I saw these two, that could be used for the normal on/off option. I don't know how complicated it is to add an icon from fontawesome. You decide.

https://fontawesome.com/v4/icon/toggle-on
https://fontawesome.com/v4/icon/toggle-off

ArneVogel commented:
I added the icons from the image: https://listudy.org/en/icons

I personally like both the toggle on off icons so I would prefer to keep them.

What icon do you have in mind for play against stockfish? I like the one currently used.

As to the pr: I would prefer to have them separate.

olleeriksson commented:
Alright. Sounds good. I thought the computer screen icon second from top left would be good for Play against Stockfish, but if you prefere the current one I'll leave it at that. And then I thought the top left icon for Analysis board, or maybe the lower study hat thing. I had some ideas to include a link to the Study in Lichess if you had created your study from one in Lichess, but then I kind of gave up on that idea. Unless you have a preference for an Analysis board icon I'll pick either the top left one or the lower bottom one depending how they feel.
